Coin Selection Better Coverage + Fix Fee Calculation

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have improved coin selection unit tests by also specifying and checking other parts of a coin selection result (i.e. outputs and change)
- [x] I have added some edge-case unit tests to test coin overflow in some cases
- [x] I have fixed a wrong strict inequality sign (caught by updated tests)
- [x] I have fixed a bug in the fee estimation were we were counting the same bytes twice.

# Comments

<!-- Additional comments or screenshots to attach if any -->

@piotr-iohk This is the best I can do in terms of coverage in a reasonable amount of time. There's still one particular overflow case that is hard to instrument in practice (impossible to meet :thinking: ?) but that is there in the code as a safe "gatekeeper" I suppose.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

parsonsmatt commented on src/Cardano/Wallet/API/Internal/Handlers.hs:
```diff
@@ -48,21 +45,6 @@ applyUpdate nc = do
         Right () ->
             pure NoContent
 
-handleNodeError :: ClientError () -> Handler a
```
Relocated into `NodeProxy` module.

parsonsmatt commented on src/Cardano/Wallet/API/V1/Handlers.hs:
```diff
@@ -11,16 +11,18 @@ import qualified Cardano.Wallet.API.V1.Handlers.Settings as Settings
+    :<|> Accounts.handlers     passiveWallet
+    :<|> Transactions.handlers w
+    :<|> Settings.handlers     nodeClient
+    :<|> Info.handlers         nodeClient
```
tempted to bite the bullet and just remove all the vertical alignment here

KtorZ replied:
I know that feel. Maybe we should allow one day per month called "alignment catharsis day" where we only submit PRs to fix variable-length vertical alignments.

KtorZ committed:
also check for change and output in the coin selection tests

KtorZ committed:
fix wrong strict inequality '>' --> '>=' in mkChange invariant

KtorZ committed:
Fix estimation of the address payload

Turns out the CBOR surrounding encoding is actually already included in
the payload size! So we don't need to count it twice.

KtorZ committed:
add edge-case test cases for coin overflow when computing change outputs

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -421,10 +419,7 @@ estimateFee policy (CoinSelection inps outs chngs) =
-        + BS.length bytes
-        + sizeOf (CBOR.encodeWord32 $ crc32 bytes)
-        + sizeOfCoin c
+        1 + BS.length bytes + sizeOfCoin c
```
So, turns out those 5 bytes + the crc are already part of the `bytes` payload! So we don't need to count them twice!

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:
```diff
@@ -377,15 +408,17 @@ propFeeEstimation (ShowFmt sel, InfiniteList chngAddrs _) =
     in
-        property (calcFee >= realFeeInf && calcFee <= realFeeSup)
+        (calcFee >= realFeeInf && calcFee <= realFeeSup, encodedTx)
+            === (True, encodedTx)
```
No change in the test logic here, but we'll now display the full transaction binary representation upon failure which is quite handy for debugging!

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:
```diff
@@ -524,8 +557,8 @@ instance Arbitrary Address where
-        mainnetA <- genAddress (39, 43)
-        testnetA <- genAddress (46, 50)
+        mainnetA <- genAddress (33, 33)
+        testnetA <- genAddress (40, 40)
```
This ought to be only the internal payload of the address. So, same confusion as within the fee calculation but, on the generation this time. The error was hard to catch because this one actually compensate for the previous one in the implementation. So, only in rare cases (1 tests out of 33k!), the double-counting would not be compensated by the wrong generation here and trigger an error!

piotr-iohk commented on test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:
```diff
@@ -299,7 +300,7 @@ spec = do
     describe "Fee Estimation properties" $ do
         it "Estimated fee is the same as taken by encodeSignedTx"
-            (withMaxSuccess 1000 $ property propFeeEstimation)
+            (withMaxSuccess 2500 $ property propFeeEstimation)
```
more the merrier :) :+1:

piotr-iohk reviewed (approved):
lgtm :+1:
